kafka-producers/templates/deployment.yaml

```diff
@@ -0,0 +1,68 @@
+{{- range $producer, $values := .Values.producers }}
+---
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: {{ printf "%s-%s" $.Release.Name $producer | trunc 63 | trimSuffix "-" }}
+  labels:
+    helm.sh/chart: {{ include "kafka-producers.chart" $ }}
+    app.kubernetes.io/instance: {{ $.Release.Name }}
+  {{- include "multicast.annotation" . | nindent 2 }}
+spec:
+  replicas: 1
+  selector:
+    matchLabels:
+      app: {{ printf "%s-%s" $.Release.Name $producer | trunc 63 | trimSuffix "-" }}
+      app.kubernetes.io/instance: {{ $.Release.Name }}
+  template:
+    metadata:
+      labels:
+        app: {{ printf "%s-%s" $.Release.Name $producer | trunc 63 | trimSuffix "-" }}
+        app.kubernetes.io/instance: {{ $.Release.Name }}
+      {{- include "multicast.annotation" . | nindent 6 }}
+    spec:
+      containers:
+        - name: {{ printf "%s-%s" $.Release.Name $producer | trunc 63 | trimSuffix "-" }}
+          {{- if .image }}
+          {{- $image := .image.repository | default $.Values.defaults.image.repository }}
+          {{- $tag := .image.tag | default $.Values.defaults.image.tag}}
```

Comment:
this pattern seems repeated (using a Values.defaults section) - why not just have the default be image: tag: in the values?  Then you could get rid of the default part - helm automatically starts off with the values.yaml and then folds in the other new values files on top of it.  Or is there something you're trying to have unset somewhere?  I think getting rid of the defaults stuff would make this even simpler, but getting down to one file is great!

Reply:
Part of the complication comes from the requirement of being able to change things per producer. Since the override comes inside the producers.name key and may not always be present, this results in all the weird checking. Helm templating doesn't have good optional chaining and requires checking key existence before using them, hence I have all the if statements.  I will drop the defaults key since I agree that's a bit redundant.